README.md
# flutter_inner_drawer
[![pub package](https://img.shields.io/badge/pub-1.0.0+1-orange.svg)](https://pub.dartlang.org/packages/flutter_inner_drawer)
[![Awesome Flutter](https://img.shields.io/badge/Awesome-Flutter-blue.svg?longCache=true&style=flat-square)](https://github.com/Solido/awesome-flutter#drawers)
[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.me/dnag88)


Inner Drawer is an easy way to create an internal side section (left/right) where you can insert a list menu or other.

## Installing
Add this to your package's pubspec.yaml file:
```dart
dependencies:
  flutter_inner_drawer: "^1.0.0+1"
```
## Demo
<div align="center">
  <table><tr>
 <td style="text-align:center">
  <img width="250px"  src="https://github.com/Dn-a/flutter_inner_drawer/raw/master/assets/img/example5.1.gif?" />
 </td>
 <td style="text-align:center">
   <img width="250px"  src="https://github.com/Dn-a/flutter_inner_drawer/raw/master/assets/img/example5.3.gif?" />
 </td>
 </tr></table>
</div>


### Simple usage
```dart
import 'package:flutter_inner_drawer/inner_drawer.dart';
.
.
.
   
    @override
    Widget build(BuildContext context)
    {
        return InnerDrawer(
            key: _innerDrawerKey,
            onTapClose: true, // default false
            swipe: true, // default true            
            colorTransitionChild: Color.red, // default Color.black54
            colorTransitionScaffold: Color.black54, // default Color.black54
            
            //When setting the vertical offset, be sure to use only top or bottom
            offset: IDOffset.only(bottom: 0.05, right: 0.0, left: 0.0),
            
            scale: IDOffset.horizontal( 0.8 ), // set the offset in both directions
            
            proportionalChildArea : true, // default true
            borderRadius: 50, // default 0
            leftAnimationType: InnerDrawerAnimation.static, // default static
            rightAnimationType: InnerDrawerAnimation.quadratic,
            backgroundDecoration: BoxDecoration(color: Colors.red ), // default  Theme.of(context).backgroundColor
            
            //when a pointer that is in contact with the screen and moves to the right or left            
            onDragUpdate: (double val, InnerDrawerDirection direction) {
                // return values between 1 and 0
                print(val);
                // check if the swipe is to the right or to the left
                print(direction==InnerDrawerDirection.start);
            },
            
            innerDrawerCallback: (a) => print(a), // return  true (open) or false (close)
            leftChild: Container(), // required if rightChild is not set
            rightChild: Container(), // required if leftChild is not set
            
            //  A Scaffold is generally used but you are free to use other widgets
            // Note: use "automaticallyImplyLeading: false" if you do not personalize "leading" of Bar
            scaffold: Scaffold(
                appBar: AppBar(
                    automaticallyImplyLeading: false
                ),
            ) 
            /* OR
            CupertinoPageScaffold(                
                navigationBar: CupertinoNavigationBar(
                    automaticallyImplyLeading: false
                ),
            ), 
            */
        )
    }
    
    //  Current State of InnerDrawerState
    final GlobalKey<InnerDrawerState> _innerDrawerKey = GlobalKey<InnerDrawerState>();    

    void _toggle()
    {
       _innerDrawerKey.currentState.toggle(
       // direction is optional 
       // if not set, the last direction will be used
       //InnerDrawerDirection.start OR InnerDrawerDirection.end                        
        direction: InnerDrawerDirection.end 
       );
    }
    
```

### InnerDrawer Parameters
|PropName|Description|default value|
|:-------|:----------|:------------|
|`scaffold`|*A Scaffold is generally used but you are free to use other widgets*|required|
|`leftChild`|*Inner Widget*|required if rightChild is not set|
|`rightChild`|*Inner Widget*|required if leftChild is not set|
|`leftOffset(deprecated)`|*Offset drawer width*|0.4|
|`rightOffset(deprecated)`|*Offset drawer width*|0.4|
|`leftScale(deprecated)`|*Left scaffold scaling*|1|
|`rightScale(deprecated)`|*Right scaffold scaling*|1|
|`offset`|*Offset InnerDrawer width*|IDOffset.horizontal(0.4)|
|`scale`|*Scaffold scaling*|IDOffset.horizontal(1)|
|`proportionalChildArea`|*If true, dynamically sets the width based on the selected offset, otherwise it leaves the width at 100% of the screen.*|true|
|`borderRadius`|*For scaffold border*|0|
|`onTapClose`|*Tap on the Scaffold closes it*|false|
|`swipe`|*activate or deactivate the swipe*|true|
|`swipeChild`|*activate or deactivate the swipeChild*|false|
|`duration`|*Animation Controller duration*|Duration(milliseconds: 246)|
|`velocity`|*Allows you to set the opening and closing velocity when using the open/close methods*|1|
|`tapScaffoldEnabled`|*Possibility to tap the scaffold even when open*|false|
|`boxShadow`|*BoxShadow of scaffold opened*|[BoxShadow(color: Colors.black.withOpacity(0.5),blurRadius: 5)]|
|`colorTransitionChild`|*Change background color while swiping*|Colors.black54|
`colorTransitionScaffold`|*Change background color while swiping*|Colors.black54|
|`leftAnimationType`|*static / linear / quadratic*|static|
|`rightAnimationType`|*static / linear / quadratic*|static|
|`backgroundDecoration`|*possibility to manage the main background Decoration*|BoxDecoration(color: Theme.of(context).backgroundColor)|
|`innerDrawerCallback`|*Optional callback that is called when a InnerDrawer is opened or closed*||
|`onDragUpdate`|*When a pointer that is in contact with the screen and moves to the right or left*||
|`_innerDrawerKey.currentState.open`|*Current State of GlobalKey<InnerDrawerState>(check example) - OPEN*||
|`_innerDrawerKey.currentState.close`|*Current State of GlobalKey<InnerDrawerState>(check example) - CLOSE*||
|`_innerDrawerKey.currentState.toggle`|*Current State of GlobalKey<InnerDrawerState>(check example) - OPEN or CLOSE*||

## Donate
It takes time to carry on this project. If you found it useful or learned something from the source code please consider the idea of donating 5, 20, 50 € or whatever you can to support the project. 
- [![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.me/dnag88)

## Issues
If you encounter problems, open an issue. Pull request are also welcome.


CHANGELOG.md
## [1.0.0+1] - 2021-03-17.

* [Fixed] - Improvement null safety support.

## [1.0.0] - 2021-03-17.

* [Updated] - Migrate to stable null safety.

## [0.5.8] - 2021-03-17.

* [Updated] - Flutter 2.0 support.
* [Updated] - example 1.

## [0.5.7+2] - 2020-08-05.

* [Clean] - deprecated functions.
* [Updated] - example 1.
* [Fixed] - issues #43, #47.

## [0.5.7+1] - 2020-07-31.

* [Updated] - Readme.

## [0.5.7] - 2020-07-31.

* [Added] - swipeChild which allows swiping scaffold from left/right child.
* [Fixed] - issue #45.
* [Replaced] - `boxDecoration` with `decoration`.
* [Updated] - Readme.

## [0.5.6+1] - 2020-05-14.

* Update Documentation.
* general improvements.
* `backgroundColor` field replaced by `backgroundDecoration` (BoxDecoration).

## [0.5.6] - 2020-05-13.

* Update Documentation.
* Added `velocity` field. Allows you to set the opening and closing velocity when using the open/close methods.
* `colorTransition` field renamed to `colorTransitionChild`.
* Added `colorTransitionScaffold` field.

## [0.5.5+3] - 2020-04-29.

* Rollback issue #33.
* Fixed some problems.

## [0.5.5+2] - 2020-04-28.

* general improvements.

## [0.5.5+1] - 2020-04-28.

* Fixed issues : #30 and #33.

## [0.5.5] - 2020-03-07.

* Update Documentation.

## [0.5.4] - 2019-12-07.

* Update Readme.

## [0.5.3] - 2019-12-07.

* **new Features and Documentation Updates**
*
* Added `IDOffset` class. An immutable set of offset in each of the four cardinal directions. 
* Added `offset` field. `leftOffset` and `rightOffset` Deprecated.
* Added `scale` field. `leftScale` and `rightScale` Deprecated.
* Added `duration` field. Duration of AnimationController.
* Cleaning code.
* Update Readme.

## [0.5.2] - 2019-11-15.

* **General improvement of the code**
*
* Added `backgroundColor` field.
* Fixed some problems related to swipe and scaleFactor.
* Cleaning code.
* Update Readme.

## [0.5.1] - 2019-11-09.

* **General improvement of the code**
*
* Added `proportionalChildArea`. if == true dynamically sets the width based on the selected offset, otherwise it leaves the width at 100% of the screen.
* Possibility to set `boxShadow` also for linear animation.
* When `borderSide` > 0, `boxShadow` did not work.
* Cleaning code.
* Update Readme.

## [0.5.0] - 2019-10-16.

* **new Features and Documentation Updates**
*
* Added `borderRadius` field for scaffold border - (double value).
* Added `leftScale` and  `RightScale` fields for scaffold scaling - (double value).
* Added `onDragUpdate(double value, InnerDraweDirection direction)` callback function.

## [0.4.0] - 2019-08-07.

* **new Features and Documentation Updates**
*
* Parameter `position` removed. Now it is possible to define `leftChild` and `rightChild` simultaneously.
* Parameter `offset` replaced with `leftOffset` and `rightOffset`.
* Parameter `animationType` replaced with `leftAnimationType` and `rightAnimationType`.
* Possibility to tap the scaffold even when open with `tapScaffoldEnabled`. 

## [0.3.0] - 2019-07-07.

* General improvement of the code.

## [0.2.9] - 2019-06-20.

* Updated dependencies.
* Added toggle method.

## [0.2.8] - 2019-06-14.

* fixed history.

## [0.2.7] - 2019-06-14.

* Cleaning code.
* Update Readme.

## [0.2.6] - 2019-05-13.

* Fixed InnerDrawerCallback.

## [0.2.5] - 2019-04-19.

* General improvement of the code.

## [0.2.4] - 2019-04-16.

* Fixed swipe precision.

## [0.2.3] - 2019-04-06.

* fixed some problem.
* General improvement of the code.

## [0.2.2] - 2019-02-26.

* fixed some artifacts with linear animation.

## [0.2.1] - 2019-02-25.

* solved the problem of CupertinoThemeData that launched an assert.

## [0.2.0] - 2019-02-16.

* 3 types of animation (static - linear - quadratic) 
* Improved documentation.

## [0.1.5] - 2019-02-13.

* Improved documentation.

## [0.1.4] - 2019-02-13.

* Improved documentation.

## [0.1.3] - 2019-02-13.

* fixed swipe.

## [0.1.2] - 2019-02-13.

* Added side trigger - Possibility to activate/deactivate the swipe.

## [0.1.1] - 2019-02-12.

* Improved documentation.

## [0.1.0] - 2019-02-12.

* Improved documentation - General improvement of the code.

## [0.0.1] - 2019-02-12.

* Created Inner Drawer.


example/lib/children/left_child_1.dart
import 'dart:ui';
import 'package:example/notifier/drawer_notifier.dart';
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

class LeftChild extends StatelessWidget {
  LeftChild({Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    print("build left");

    final double swipeOffset =
        context.select((DrawerNotifier value) => value.swipeOffset);
    return Material(
        child: Stack(
      children: <Widget>[
        Container(
          decoration: BoxDecoration(
            gradient: LinearGradient(
              begin: Alignment.topRight,
              end: Alignment.bottomLeft,
              // Add one stop for each color. Stops should increase from 0 to 1
              //stops: [0.1, 0.5,0.5, 0.7, 0.9],
              colors: [
                ColorTween(
                  begin: Colors.blueAccent,
                  end: Colors.blueGrey[400].withRed(100),
                ).lerp(swipeOffset),
                ColorTween(
                  begin: Colors.green,
                  end: Colors.blueGrey[800].withGreen(80),
                ).lerp(swipeOffset),
              ],
            ),
          ),
          child: Stack(
            children: <Widget>[
              Padding(
                padding: EdgeInsets.only(left: 30),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: <Widget>[
                    Column(
                      children: <Widget>[
                        Container(
                          margin: EdgeInsets.only(left: 10, bottom: 15),
                          width: 80,
                          child: ClipRRect(
                            child: Image.network(
                              "https://img.icons8.com/officel/2x/user.png",
                            ),
                            borderRadius: BorderRadius.circular(50),
                          ),
                        ),
                        Text(
                          "User",
                          style: TextStyle(color: Colors.white, fontSize: 18),
                        )
                      ],
                      //mainAxisAlignment: MainAxisAlignment.center,
                    ),
                    Padding(
                      padding: EdgeInsets.all(10),
                    ),
                    ListTile(
                      onTap: () => print("Dashboard"),
                      title: Text(
                        "Dashboard",
                        style: TextStyle(color: Colors.white, fontSize: 14),
                      ),
                      leading: Icon(
                        Icons.dashboard,
                        color: Colors.white,
                        size: 22,
                      ),
                    ),
                    ListTile(
                      title: Text(
                        "Nametag",
                        style: TextStyle(fontSize: 14),
                      ),
                      leading: Icon(
                        Icons.rounded_corner,
                        size: 22,
                      ),
                    ),
                    ListTile(
                      title: Text(
                        "Favorite",
                        style: TextStyle(fontSize: 14),
                      ),
                      leading: Icon(
                        Icons.bookmark_border,
                        size: 22,
                      ),
                    ),
                    ListTile(
                      title: Text(
                        "Close Friends",
                        style: TextStyle(fontSize: 14),
                      ),
                      leading: Icon(
                        Icons.list,
                        size: 22,
                      ),
                    ),
                    ListTile(
                      title: Text(
                        "Suggested People",
                        style: TextStyle(fontSize: 14),
                      ),
                      leading: Icon(
                        Icons.person_add,
                        size: 22,
                      ),
                    ),
                  ],
                ),
              ),
              Positioned(
                bottom: 0,
                child: Container(
                  alignment: Alignment.bottomLeft,
                  margin: EdgeInsets.only(top: 50),
                  padding: EdgeInsets.symmetric(vertical: 15, horizontal: 25),
                  width: double.maxFinite,
                  child: Row(
                    mainAxisAlignment: MainAxisAlignment.start,
                    children: <Widget>[
                      Icon(
                        Icons.all_out,
                        size: 18,
                        color: Colors.grey,
                      ),
                      Text(
                        " LogOut",
                        style: TextStyle(
                          fontSize: 16,
                          color: Colors.grey,
                        ),
                      ),
                    ],
                  ),
                ),
              )
            ],
          ),
        ),
        swipeOffset < 1
            ? BackdropFilter(
                filter: ImageFilter.blur(
                    sigmaX: (10 - swipeOffset * 10),
                    sigmaY: (10 - swipeOffset * 10)),
                child: Container(
                  decoration: BoxDecoration(
                    color: Colors.black.withOpacity(0),
                  ),
                ),
              )
            : null,
      ].where((a) => a != null).toList(),
    ));
  }
}
